fdb-extensions/src/main/java/com/apple/foundationdb/async/RTree.java

```diff
@@ -360,8 +482,36 @@ public Config getConfig() {
     @Nonnull
     public AsyncIterator<ItemSlot> scan(@Nonnull final ReadTransaction readTransaction,
                                         @Nonnull final Predicate<Rectangle> mbrPredicate) {
-        AsyncIterator<LeafNode> leafIterator = new LeafIterator(readTransaction, rootId, mbrPredicate);
-        return new ItemIterator(leafIterator);
+        return scan(readTransaction, null, null, mbrPredicate);
+    }
+
+    /**
+     * Perform a scan over the tree within the transaction passed in using a predicate that is also passed in to
+     * eliminate subtrees from the scan. This predicate may be stateful which allows for dynamic adjustments of the
+     * queried area while the scan is active.
+     * <br>
+     * A scan of the tree offers all items that pass the {@code mbrPredicate} test in Hilbert Value order using an
+     * {@link AsyncIterator}. The predicate that is passed in is applied to intermediate nodes as well as leaf nodes,
+     * but not to elements contained in a leaf node. The caller should filter out items in a downstream operation.
+     * A scan of the tree will not prefetch the next node before the items of the current node have been consumed. This
+     * guarantees that the semantics of the mbr predicate can be adapted in response to the items being consumed.
+     * (this allows for efficient scans for {@code ORDER BY x, y LIMIT n} queries).
```

Comment:
I'm not sure I follow how the MBR-predicate can be massaged in a way to efficiently execute "order by x, y limit n" queries. Is this alluding to some kind of scan that:

1. Keeps a "top n" priority queue of x-y and
2. Adjusts the MBR to account for the elements it's already seen (e.g., excluding values outside the already seen top `n` x values)

I guess that's cool that the predicate can be adjusted to do that. It does seem a bit weird to me that the item slots aren't also subject to the mbrPredicate, though

Reply:
Yes, we do keep a queue of size `n` and then adjust the predicate to cut out areas that we definitely don't have to visit. See `RTreeScanTest.queryTopNWithFilters` for the actual implementation. The gist is that you read your top `n` records and look at the `xn, yn` coordinate. You now know that you never have to look `x > xn` and can slice a piece off the search rectangle.

Reply:
Wrt to your last sentence in the comment. A leaf node has an `mbr` whereas an item within a leaf node only has a position. The `LeafNodeIterator` advances you from leaf node to leaf node in Hilbert values order hopefully skipping many leaf nodes that won't contain any useful data pertaining to the query at hand. The leaf nodes presented to you in the iterator contain a superset of items you are interested in. The check on item level is done using just an ordinary async filter on the item slot iterator that encapsulates a leaf node iterator. That filter checks for containment of the items position, not for the overlapping of a search region with an mbr. I hope this makes sense.